src/devices/Pca95x4/Pca95x4.cs

```diff
@@ -0,0 +1,199 @@
+            }
+        }
+
+        private static void ClearBit(ref byte data, int bitNumber)
+        {
+            ValidateBitNumber(bitNumber);
+            data &= (byte)~(1 << bitNumber);
+        }
+
+        private void SetBit(ref byte data, int bitNumber)
+        {
+            ValidateBitNumber(bitNumber);
+            data |= (byte)(1 << bitNumber);
+        }
+
+        private static bool GetBit(byte data, int bitNumber)
+        {
+            ValidateBitNumber(bitNumber);
+            return ((data >> bitNumber) & 1) == 1;
+        }
+
+        private void InitializeMasterGpioController()
+        {
+            // Only need master controller if there is external pin provided.
+            if (_interrupt != null)
+            {
+                _masterGpioController = new GpioController();
+                _masterGpioController.OpenPin((int)_interrupt, PinMode.Input);
+            }
+        }
+
+        /// <summary>
+        /// Reads a byte from a register.
+        /// </summary>
+        /// <param name="register">The register to read.</param>
+        /// <returns>The data read from the register.</returns>
+        public byte Read(Register register)
+        {
+            byte[] data = Read(register, 1);
+            return data[0];
+        }
+
+        /// <summary>
+        /// Reads a number of bytes from a register. 
+        /// All data returned is from the selected register as Pca95x4 does not auto-increment addressing.
+        /// </summary>
+        /// <param name="register">The register to read.</param>
+        /// <param name="byteCount"></param>
+        /// <returns>The data read from the register.</returns>
+        public byte[] Read(Register register, byte byteCount)
+        {
+            Write(register, new byte[] { }); // Set address to register first.
+
+            byte[] readBuffer = new byte[byteCount];
+            _i2cDevice.Read(readBuffer);
+            return readBuffer;
+        }
+
+        /// <summary>
+        /// Reads a bit from a register.
+        /// </summary>
+        /// <param name="register">The register to read.</param>
+        /// <param name="bitNumber">The register bit number to read.</param>
+        /// <returns>The value of the register bit read.</returns>
+        public bool ReadBit(Register register, int bitNumber)
+        {
+            byte data = Read(register);
+            return GetBit(data, bitNumber);
+        }
+
+        /// <summary>
+        /// Reads the pin value of the interrupt (INT).
+        /// </summary>
+        /// <returns>The pin value of the interrupt (INT).</returns>
+        public PinValue ReadInterrupt()
+        {
+            if (_masterGpioController == null)
+            {
+                throw new Exception("Master controller has not been initialized.");
+            }
+
+            if (_interrupt == null)
+            {
+                throw new Exception("INT pin has not been initialized.");
```

Comment:
correct me if I'm wrong but I believe we should never come to this line right? since if you didn't have an interrupt you probably already fell into the previous if check.

Reply:
I'm not sure if I understand the question.  Do you mean reach the code at  `if (_interrupt == null)` as it would throw at `if (_masterGpioController == null)`?  If so, then you are right in this example.

The code was sort of a copy from Mcp23xxx, but the difference is the Mcp23xxx has a few other pins it offers.  There is only the interrupt in this binding.  Therefore, I guess you could do the check if the masterController was null and wouldn't need to do the check if pin was null in this case.

This particular function has nothing do with a real interrupt per se.  It was more of a helper to read the particular pin if user wanted.  I still need to study on GPIO events and then I was going to wire up to the interrupt pin at some point.

So are you asking to remove `if (_interrupt == null)`?  If so, I'm good with that in this case and can do.

Reply:
Yes, I was referring that if _interrupt was null, then you would have already fallen inside the if of _masterGpioController. I think it is just code that at least with the way you have the code written right now, would never get executed, which is why I wouls just suggest to remove it.